staging/src/k8s.io/apiserver/pkg/storage/names/generate_test.go

```diff
@@ -19,11 +19,57 @@ package names
 import (
 	"strings"
 	"testing"
+
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
 )
 
 func TestSimpleNameGenerator(t *testing.T) {
-	name := SimpleNameGenerator.GenerateName("foo")
-	if !strings.HasPrefix(name, "foo") || name == "foo" {
-		t.Errorf("unexpected name: %s", name)
+	mkBase := func(n int) string {
+		return utilrand.String(n-1) + "-"
+	}
+
+	cases := []struct {
+		base        string
+		expectLen   int
+		expectTrunc int
+	}{{
+		// Normal use
+		base:      mkBase(3),
+		expectLen: 3 + randomLengthMax,
+	}, {
+		// Edge case: long name, just enough for max suffix size
```

Comment:
I think this will break nested name generation, eg deployment makes RS makes pods; we want the pods to get a long suffix but pods that work today might stop working after this because the RS name takes up 5 more characters.

Reply:
RS doesn't use `generateName`:

pkg/controller/deployment/sync.go
```
    // Create new ReplicaSet
    newRS := apps.ReplicaSet{
        ObjectMeta: metav1.ObjectMeta{
            // Make the name deterministic, to ensure idempotence
            Name:            d.Name + "-" + podTemplateSpecHash,                                                                                                                     
            Namespace:       d.Namespace,
```

This is, obviously, broken.  Deployment name is validated with `NameIsDNSSubdomain`.  So is ReplicaSet.
So if Deployment.name is too long, ReplicSet creation just fails.  https://github.com/kubernetes/kubernetes/issues/116447

But let's assume that was fixed.  :) 

No matter what deployment name we feed it, we get a replicaset with a 10 character suffix (well, 10 + '-'), potentially at the cost of the last few characters of the deployment name.

Then, no matter what replicaset name we feed into `generateName`, we get a pod with at least 5 characters of suffix, again at the expense of the last few characters of input.

So what pods would fail?  We're only using extra space *if* we can do so and stay less than 63.  We don't truncate any extra.  Worst case is exactly the same as today, unless I am missing something?

Reply:
ah, if RS doesn't use generateName then it won't have the problem I described. Do we have any other "thingy makes thingy makes thingy" relationships?

Do we think CR / operator authors will do that very often?

Reply:
There are some (like Service will eventually create an IP resource, but they are 1:1, not 1:N).  I can't think off the top of my head of anything that "thingy makes collection-of-thingy".  And if we did they are all either self-truncating or subject to the same "too long" GIGO problem.

Reply:
I take that back - Service -> EndpointSlice is an example, but service names are short (DNS Label) and EPS names are long (DNS Subdomain).

Reply:
It's only a problem if the output of generateName is used in a subsequent generateName. I don't think the those examples do that? But CR / operator authors might be doing it :/

Reply:
Here's a crazy idea to fix this:

What if generate name detected names that contain a generate-name produced suffix and truncated that part of the name before adding on a new suffix?

Reply:
But that's a problem today, this just makes it 5 chars sooner.

E.g. A -> B -> C.

len(A.name) is 63
len(B.generateName) = 63 -> B.name = A[0:58]+rand(5)
len(C.generateName) = 63 -> C.name = A[0:58]+rand(5)

It's unfortunate that C lost its name-link to B, but it should still be reasonably-likely unique.

Unless a C is always derived from a B, it doesn't know to do middle-out truncation.  In hindsight maybe we should have tracked the original name + generated portion for this use-case, but...not all problems are worth fixing?

Reply:
> What if generate name detected names that contain a generate-name produced suffix 

Why?  The only property we need to (approximately) ensure is uniquess.  I doubt it's worth setting up something so easy to trip over - what's the ROI ?

Reply:
argh, I just realized the existing and proposed code both truncate the base name if it's too long, so it doesn't have the problem I thought it did.

Truncating the base name potentially increases the number of things we can collide with. I'd vote for truncating it by more than just 5 characters if we need to truncate.

Reply:
We can do that - I can just straight change 5 to 10, nothing will break.  Mostnames will just get 5 charts longer.  Long names will get a little less prefix and a little more suffix.

I figured we're getting away with 5, so why churn if not needed, but that's such an edge case.  It is simpler just to go big.  We could do 8 instead of 10, too.  Still pretty unique :)